### PR TITLE
Improve structure of core and plugins for lib and cli 

### PR DESCRIPTION
**Pull request by ZzAve**
WIP

Conceptual hierarchy:

- cli dependsOn lib
- other plugins dependOn lib
- lib depends on core

## Lib
The `lib` module exposed a lib for all platforms.
This includes an npm module (that does not offer cli integration)

## Cli
The `cli` module exposes executables for all platforms, native, js, and jvm(?)
This includes an npm module that offers cli integration, and includes / depends on the other npm module.

Checks:

- [ ] Test integration with intellij plugin
- [ ] Test integration with vscode plugin
- [ ] Test npm integration locally
- [ ] Add tests
- [ ] Add documentation about module structure

### CODE REVIEW

File: src/compiler/lib/build.gradle.kts [hunk 1]
```diff
@@ -40,12 +39,43 @@ kotlin {
...
+        }
+    }
+}
+
+fun String.fromEnv() = let(System::getenv).toBoolean()
+
+fun KotlinNativeTargetWithHostTests.build() = binaries {
+    staticLib {
+    }
+}
```
Comment (ZzAve) [inline src/compiler/lib/build.gradle.kts:71]: Expose native libs in the `lib` module (this sounds like a no brainer)

File: src/compiler/lib/build.gradle.kts [hunk 2]
```diff
@@ -40,12 +39,43 @@ kotlin {
...
+
+fun KotlinJsTargetDsl.build() {
+    nodejs()
+//        generateTypeScriptDefinitions()
+    binaries.library()
+    compilations["main"].packageJson {
+        customField("name", "@flock/wirespec")
+        customField("bin", mapOf("wirespec" to "wirespec-bin.js"))
     }
 }
```
Comment (ZzAve) [inline src/compiler/lib/build.gradle.kts:81]: Expose a `js` lib in the `lib` module. This does not include cli logic.

File: src/plugin/cli/build.gradle.kts [hunk 3]
```diff
@@ -63,9 +63,8 @@ kotlin {
     sourceSets {
         val commonMain by getting {
             dependencies {
+                implementation(project(":src:compiler:lib"))
                 implementation(project(":src:plugin:arguments"))
-                implementation(project(":src:compiler:core"))
-                implementation(project(":src:converter:openapi"))
                 implementation(CLI_LIB)
             }
         }
```
Comment (ZzAve) [inline src/plugin/cli/build.gradle.kts:70]: `cli` depends on the `lib` module, using the libraries exposed for all platforms. This alleviates the 'weird' need for `jsMain` only to depend on the `lib` module.

File: src/plugin/cli/src/jsMain/kotlin/community/flock/wirespec/plugin/cli/Cli.kt [hunk 4]
```diff
@@ -15,6 +15,8 @@ fun cli(args: Array<String>) {
     main(args)
 }
 
+
+// FIXME: remove me? Should be part of cli?
 @JsExport
 fun parser(source: String): Array<WsNode> {
```
Comment (ZzAve) [inline src/plugin/cli/src/jsMain/kotlin/community/flock/wirespec/plugin/cli/Cli.kt:21]: Does the CLI need to expose this? Shouldn't this be part of the lib instead?

**Comment by wilmveel** [comment]
From a code organisation perspective I get why it is better to split the cli and lib into two package. 

From a consumer perspective I think it is beter to have one npm package published (@flock/wirespec).

- One package is simple to document and understand.
- With two packages versions can go out of sync wich can give unexpected behavior.
- The code in both packages will be almost the same
- Package size is not an issue they are relative small and will be tree shaked when bundled with an app.

**Comment by ZzAve** [comment]
> From a code organisation perspective I get why it is better to split the cli and lib into two package.
> 
> From a consumer perspective I think it is beter to have one npm package published (@flock/wirespec).
> 
> * One package is simple to document and understand.
> * With two packages versions can go out of sync wich can give unexpected behavior.
> * The code in both packages will be almost the same
> * Package size is not an issue they are relative small and will be tree shaked when bundled with an app.

We could make the `cli` js lib expose both, and not expose an npm module from the lib folder, maybe? That way, we can still expose libs in the lib (mostly maven/gradle deps), and consider npm modules and native clis the other module. If need be we could have an `npm` module, and a `cli` module, to split stuff, te ensure we can generate the typescript definitions.


core -> core
lib  -> maven/gradle libs

cli -> native clis 
npm -> npm module (non multplatform)

**Comment by wilmveel** [comment]
I have tried to expose the lib via the cli module. Then the typescript issue occurred. That is why the npm module was introduced. This module only has a js source set. Maybe the location of the module in plugins is a bit wierd. To me it makes sense to have one module to control what is published to npm

**Pull request closed without merging**